### PR TITLE
:umbrella: Firm integration

### PR DESCRIPTION
**Pull request by I-Al-Istannen**
## Todo:
- [x] While loop
- [x] malloc for new expr
- [x] stdlib for print/write/flush
- [x] implicit return 
- [x] Invert cmp if negated
- [x] Think about maturing
- [x] Continuation passing style emulation in visitor?
- [x] Segfault in JVM
- [x] Name mangling
- [x] Exceptions
- [x] Array größe 0
- [x] [Cond/Expr in getrennte Methoden, explizite Duplikate wo nötig]?
- [x] calloc
- [x] Extract `runtime.c` at runtime from resources folder

### CODE REVIEW

File: src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import firm.MethodType;
+import firm.Type;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class EntityHelper {
```
Comment (chrisliebaer) [inline src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java:19]: Some documentation would be nice.

File: src/main/java/com/github/firmwehr/gentle/firm/construction/SlotTable.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import com.github.firmwehr.gentle.semantic.ast.LocalVariableDeclaration;
+import com.github.firmwehr.gentle.semantic.ast.SClassDeclaration;
+import com.github.firmwehr.gentle.semantic.ast.SMethod;
+import com.github.firmwehr.gentle.semantic.ast.expression.SLocalVariableExpression;
+import com.github.firmwehr.gentle.source.SourceSpan;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SlotTable {
```
Comment (chrisliebaer) [inline src/main/java/com/github/firmwehr/gentle/firm/construction/SlotTable.java:16]: Javadoc

File: src/main/java/com/github/firmwehr/gentle/firm/construction/TypeHelper.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import firm.ClassType;
+import firm.Mode;
+import firm.PointerType;
+import firm.PrimitiveType;
+import firm.Type;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class TypeHelper {
```
Comment (chrisliebaer) [inline src/main/java/com/github/firmwehr/gentle/firm/construction/TypeHelper.java:23]: Javadoc

**Review by chrisliebaer — CHANGES_REQUESTED** [review]

**Review by chrisliebaer — APPROVED** [review]

**Review by pbrinkmeier — COMMENTED** [review]
Inspected everything but FirmGraphBuilder, seems good.

**Comment by pbrinkmeier** [comment]
I will also look through FirmGraphBuilder and tonight we can merge :)

File: src/main/java/com/github/firmwehr/gentle/firm/construction/SlotTable.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		return new SlotTable(map);
+	}
+
+	private SlotTable(Map<LocalVariableDeclaration, Integer> map) {
+		this.toIndexMap = map;
+	}
+
+	public int computeIndex(LocalVariableDeclaration localVariable) {
+		return Optional.ofNullable(this.toIndexMap.get(localVariable)).orElseThrow(() -> {
+			throw new InternalCompilerException("encountered unknown variable " + localVariable);
```
Comment (Garmelon) [inline src/main/java/com/github/firmwehr/gentle/firm/construction/SlotTable.java:73]: Throwing exception inside `orElseThrow`

File: src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+	public Entity getEntity(SField field) {
+		return fieldEntities.get(field);
+	}
+
+	public Entity getEntity(StdLibEntity entity) {
+		return entity.getEntity();
+	}
+
+	public Entity computeMethodEntity(SMethod method) {
```
Comment (Garmelon) [inline src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java:51]: Why `computeMethodEntity` but `setFieldEntity`?

**Review by Garmelon — APPROVED** [review]

File: src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+		}
+
+		if (method.isStatic()) {
+			returnType = new Type[]{typeHelper.getType(new SIntType())};
+		}
+
+		MethodType methodType = new MethodType(types, returnType);
+		Entity entity = new Entity(ownerType, method.name().ident(), methodType);
+		if (!method.isStatic()) {
+			entity.setLdIdent(method.classDecl().name().ident() + "_" + method.name().ident());
```
[the remark this answers is not in the document]
Comment (I-Al-Istannen) [reply src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java:79]: Do we only want to support linux binaries?

File: src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java (hunk 5 above)
Comment (I-Al-Istannen) [reply src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java:51]: It returns an entity, the field one just creates one (returns void). I am open for suggestions for better names though

Comment (Garmelon) [reply src/main/java/com/github/firmwehr/gentle/firm/construction/EntityHelper.java:51]: Ah, I missed the return value

**Review by pbrinkmeier — APPROVED** [review]

**Pull request merged**